File: f#/words/test/WordsTest.fs
namespace test.WordsTests

open NUnit.Framework
open lib.Words

[<TestFixture>]
module WordsTests =

    [<Test>]
    let ``Test Reverse String`` () =
        Assert.AreEqual("olleh", Words.reverseString "hello")
        Assert.AreEqual("a", Words.reverseString "a")
        Assert.AreEqual("", Words.reverseString "")

    [<Test>]
    let ``Test Is Palindrome`` () =
        Assert.AreEqual(true, Words.isPalindrome "race car")
        Assert.AreEqual(true, Words.isPalindrome "level")
        Assert.AreEqual(false, Words.isPalindrome "hello")
        Assert.AreEqual(true, Words.isPalindrome "a")
        Assert.AreEqual(true, Words.isPalindrome "")

    [<Test>]
    let ``Test Is Substring`` () =
        Assert.AreEqual(true, Words.kmpSearch "test" "this is a test")
        Assert.AreEqual(false, Words.kmpSearch "not" "this is a test")
        Assert.AreEqual(true, Words.kmpSearch "" "any string")
        Assert.AreEqual(true, Words.kmpSearch "abc" "abc")
        Assert.AreEqual(false, Words.kmpSearch "abc" "ab")

    [<Test>]
    let ``Test Longest Common Subsequence`` () =
        Assert.AreEqual(4, Words.lcs "AGGTAB" "GXTXAYB")
        Assert.AreEqual(3, Words.lcs "ABC" "ABC")
        Assert.AreEqual(0, Words.lcs "ABC" "DEF")
        Assert.AreEqual(0, Words.lcs "" "ABC")
        Assert.AreEqual(0, Words.lcs "ABC" "")

File: _experimental/_r/words/test/words_test.R
library(testthat)

source("../src/words.R")

test_that("reverse string", {
  expect_equal(reverse_string("hello"), "olleh")
  expect_equal(reverse_string("a"), "a")
  expect_equal(reverse_string(""), "")
})

test_that("is palindrome", {
  expect_true(is_palindrome("race car"))
  expect_true(is_palindrome("level"))
  expect_false(is_palindrome("hello"))
  expect_true(is_palindrome("a"))
  expect_true(is_palindrome(""))
})

test_that("is substring", {
  expect_true(kmp_search("test", "this is a test"))
  expect_false(kmp_search("not", "this is a test"))
	expect_true(kmp_search("", "any string"))
	expect_true(kmp_search("abc", "abc"))
	expect_false(kmp_search("abc", "ab"))
})

test_that("longest common subsequence", {
  expect_equal(lcs("AGGTAB", "GXTXAYB"), 4)
	expect_equal(lcs("ABC", "ABC"), 3)
	expect_equal(lcs("abc", "def"), 0)
	expect_equal(lcs("", "abc"), 0)
	expect_equal(lcs("abc", ""), 0)
})